certora/spec/CVL2/timelockAuthorizerTed.spec

```diff
@@ -0,0 +1,556 @@
+// A granter can be added only by calling `addGranter` or `claimRoot`
+// A granter can only be removed by calling `removeGranter` or `claimRoot`.
+// https://prover.certora.com/output/40577/fd84cc7fb32d4748b84df810aa5c557d/?anonymousKey=b6913e028107a567c85b7de953db2467f1bb8f8b
+rule isGranterChangesOnlyWithAddOrRemoveGranter(env e, method f) {
+    bytes32 actionId;
+    address account;
+    address where;
+    bool isGranterBefore = isGranter(actionId, account, where);
+    address rootBefore = getRoot();
+    address pendingRootBefore = getPendingRoot();
+
+    // Invoke any function
+    calldataarg args;
+    f(e, args);
+
+    bool isGranterAfter = isGranter(actionId, account, where);
+
+    assert isGranterBefore && !isGranterAfter =>
+                f.selector == sig:removeGranter(bytes32, address, address).selector ||
+                f.selector == sig:claimRoot().selector,
+                "User ceased to be granter not with removeGranter or claimRoot.";
+    assert !isGranterBefore && isGranterAfter =>
+                f.selector == sig:addGranter(bytes32, address, address).selector ||
+                f.selector == sig:claimRoot().selector,
+                "User became granter not with addGranter or claimRoot.";
+    assert isGranterBefore != isGranterAfter =>
+        e.msg.sender == rootBefore || e.msg.sender == pendingRootBefore,
+        "Granter changed by somebody, who was not root or pending root.";
+}
+
+
+// STATUS - verified
+// https://prover.certora.com/output/40577/fd84cc7fb32d4748b84df810aa5c557d/?anonymousKey=b6913e028107a567c85b7de953db2467f1bb8f8b
+rule isRevokerChangesOnlyWithAddOrRemoveRevoker(env e, method f) {
+    address account;
+    address where;
+    bool isRevokerBefore = isRevoker(account, where);
+    address rootBefore = getRoot();
+    address pendingRootBefore = getPendingRoot();
+
+    // Invoke any function
+    calldataarg args;
+    f(e, args);
+
+    bool isRevokerAfter = isRevoker(account, where);
+
+    assert isRevokerBefore && !isRevokerAfter =>
+                f.selector == sig:removeRevoker(address, address).selector ||
+                f.selector == sig:claimRoot().selector,
+                "User ceased to be revoker not with removeRevoker or claimRoot.";
+    assert !isRevokerBefore && isRevokerAfter =>
+                f.selector == sig:addRevoker(address, address).selector ||
+                f.selector == sig:claimRoot().selector,
+                "User became revoker not with addRevoker or claimRoot.";
+    assert isRevokerBefore != isRevokerAfter =>
+        e.msg.sender == rootBefore || e.msg.sender == pendingRootBefore,
+        "Revoker changed by somebody, who was not root or pending root.";
+}
+
+
+// STATUS - verified
+// https://prover.certora.com/output/40577/fd84cc7fb32d4748b84df810aa5c557d/?anonymousKey=b6913e028107a567c85b7de953db2467f1bb8f8b
+rule isCancelerChangesOnlyWithAddOrRemoveCanceler(env e, method f) {
```

Comment:
missing description, probably there are more, but I won't mark them. I think you will find them anyway